### PR TITLE
⭐️ aws provider to v9

### PR DESCRIPTION
**Pull request by vjeffrey**
- [x] convert all aws resources
- [x] green build/test
- [x] license added

perhaps in a followup pr..
- [x] fix up for resource errors
- [ ] make sure all latest fixes from main are also applied here
- [x] credential options
- [x] discovery for all api resources
- [x] discovery for instances etc

### CODE REVIEW

**Comment by chris-rock** [comment]
we can probably reuse https://github.com/mondoohq/cnquery/pull/1544 here as well

File: providers/aws/connection/connection.go [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+		o = append(o, WithProfile(awsProfile))
+	}
+
+	if role, ok := m["role"]; ok {
+		log.Debug().Str("role", role).Msg("using aws sts assume role")
+		cfg, _ := config.LoadDefaultConfig(context.Background())
+		externalId := m["external-id"]
+		o = append(o, WithAssumeRole(cfg, role, externalId))
+	}
+	return o, m["profile"]
```
Comment (czunker) [inline providers/aws/connection/connection.go]: The profile is also part of the options. Why do we need to return it twice?

File: providers/aws/provider/provider.go [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+		Upstream:       upstream,
+	}
+
+	return conn, err
+}
+
+func (s *Service) detect(asset *inventory.Asset, conn *connection.AwsConnection) error {
+	asset.Id = conn.Conf.Type + "://" + conn.AccountId()
+	asset.Name = conn.Conf.Host
+	asset.Platform = &inventory.Platform{
```
Comment (czunker) [inline providers/aws/provider/provider.go]: Could we re-use `getPlatformForObject` here? It looks quite similar.

File: providers/aws/resources/asset.go [hunk 3]
```diff
@@ -0,0 +1,4 @@
+// Copyright (c) Mondoo, Inc.
```
Comment (czunker) [inline providers/aws/resources/asset.go]: Looks like we do not need this file.

File: providers/aws/resources/aws_acm.go [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+	mapTags := make(map[string]interface{})
+	for i := range tags {
+		if tags[i].Key != nil && tags[i].Value != nil {
+			mapTags[*tags[i].Key] = *tags[i].Value
+		}
+	}
+	return mapTags
+}
+
+// func (a *mqlAwsAcmCertificate) certificate() (*mqlCertificate, error) {
```
Comment (czunker) [inline providers/aws/resources/aws_acm.go]: Is this one of the resources which isn't working yet?

File: providers/aws/resources/aws_accessanalyzer.go [hunk 5]
```diff
@@ -70,13 +66,14 @@ func (a *mqlAwsAccessAnalyzer) getAnalyzers(provider *aws_provider.Provider) []*
...
-					mqlAnalyzer, err := a.MotorRuntime.CreateResource("aws.accessanalyzer.analyzer",
-						"arn", core.ToString(analyzer.Arn),
-						"name", core.ToString(analyzer.Name),
-						"status", string(analyzer.Status),
-						"type", string(analyzer.Type),
-						"tags", core.StrMapToInterface(analyzer.Tags),
-					)
+					mqlAnalyzer, err := a.MqlRuntime.CreateResource(a.MqlRuntime, "aws.accessanalyzer.analyzer",
+						map[string]*llx.RawData{
+							"arn":    llx.StringData(toString(analyzer.Arn)),
```
Comment (czunker) [inline providers/aws/resources/aws_accessanalyzer.go]: No need to add a new func. The `core.ToX` moved to `convert.ToX`.

You could use `convert.ToString` or `llx.StringDataPtr` here.

File: providers/aws/resources/aws_cloudtrail.go [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+				res = append(res, mqlAwsCloudtrailTrail)
+			}
+			return jobpool.JobResult(res), nil
+		}
+		tasks = append(tasks, jobpool.NewJob(f))
+	}
+	return tasks
+}
+
+func (a *mqlAwsCloudtrailTrail) s3bucket() (*mqlAwsS3Bucket, error) {
```
Comment (czunker) [inline providers/aws/resources/aws_cloudtrail.go:183]: I do not understand why we need this. I read this as the AWS object just does not have a S3 bucket. Couldn't we just remove the field from the MQL resource? The same for the next two funcs.

File: providers/aws/resources/shared.go [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+	var respErr *http.ResponseError
+	if errors.As(err, &respErr) {
+		if respErr.HTTPStatusCode() == 400 && strings.Contains(respErr.Error(), "AccessDeniedException") {
+			return true
+		}
+	}
+	return false
+}
+
+func toBool(b *bool) bool {
```
Comment (czunker) [inline providers/aws/resources/shared.go]: For all these `toX` funcs, there should already be one in the `convert` package or llx has also `XDataPtr` versions which might make it even simpler.

File: providers/aws/resources/discovery_conversion.go [hunk 8]
```diff
@@ -0,0 +1,559 @@
...
+	platformid := MondooObjectID(mqlObject.awsObject)
+	t := conn.Conf
+	t.PlatformId = platformid
+	return &inventory.Asset{
+		PlatformIds: []string{platformid, mqlObject.awsObject.arn},
+		Name:        mqlObject.name,
+		Platform: &inventory.Platform{
+			Name:    info.Name,
+			Title:   info.Title,
+			Kind:    "KIND_AWS_OBJECT",
```
Comment (czunker) [inline providers/aws/resources/discovery_conversion.go]: I think, this should be "aws-object".

That's how we translated it in v8:
```
        case Kind_KIND_AWS_OBJECT:
                return "aws-object"
```

**Review by czunker — COMMENTED** [review]
Oh wow. What a re-write:
image

Thanks @vjeffrey

**Comment by czunker** [comment]
Could we also remove the aws parts from the `_motor` directory with this PR?

File: providers/aws/gen/main.go [hunk 9]
```diff
@@ -0,0 +1,13 @@
+// Copyright (c) Mondoo, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package main
+
+import (
+	"go.mondoo.com/cnquery/providers-sdk/v1/plugin/gen"
+	"go.mondoo.com/cnquery/providers/core/config"
```
Comment (czunker) [inline providers/aws/gen/main.go]:
```suggestion
	"go.mondoo.com/cnquery/providers/aws/config"
```

File: Makefile [hunk 10]
```diff
@@ -299,6 +309,8 @@ providers/gomodtidy:
 	@$(call gomodtidyProvider, providers/gcp)
 	@$(call gomodtidyProvider, providers/k8s)
 	@$(call gomodtidyProvider, providers/azure)
+	@$(call testGpModProvider, providers/aws)
```
Comment (czunker) [inline Makefile]:
```suggestion
	@$(call gomodtidyProvider, providers/aws)
```

File: providers/aws/main.go [hunk 11]
```diff
@@ -0,0 +1,15 @@
+// Copyright (c) Mondoo, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package main
+
+import (
+	"os"
+
+	"go.mondoo.com/cnquery/providers-sdk/v1/plugin"
+	"go.mondoo.com/cnquery/providers/network/provider"
```
Comment (czunker) [inline providers/aws/main.go]:
```suggestion
	"go.mondoo.com/cnquery/providers/aws/provider"
```

File: providers/aws/resources/shared.go (hunk 7 above)
Comment (vjeffrey) [reply providers/aws/resources/shared.go]: oh nice!

File: providers/aws/resources/aws_acm.go (hunk 4 above)
Comment (vjeffrey) [reply providers/aws/resources/aws_acm.go]: ya, the acm certificate stuff -- it was relying on a resource that's available in core and i need to see how that all will work now

File: providers/aws/provider/provider.go (hunk 2 above)
Comment (vjeffrey) [reply providers/aws/provider/provider.go]: oh, ya, probably!

File: providers/aws/resources/aws_cloudtrail.go (hunk 6 above)
Comment (vjeffrey) [reply providers/aws/resources/aws_cloudtrail.go:183]: hmm, ill look

**Comment by vjeffrey** [comment]
ah dang, i hadn't tried to scan with the inventory pack, thank you, ill give that a try

File: providers/aws/connection/connection.go (hunk 1 above)
Comment (vjeffrey) [reply providers/aws/connection/connection.go]: idk why i did that, removing

File: providers/aws/resources/aws_accessanalyzer.go (hunk 5 above)
Comment (vjeffrey) [reply providers/aws/resources/aws_accessanalyzer.go]: oh nice, thank you

**Comment by vjeffrey** [comment]
@czunker fixing the first error you found led me to seeing that i'm hitting some error on the security groups resource - it seems inconsistent, i need to take a deeper look, otherwise the rest of your feedback should be addressed now

**Comment by vjeffrey** [comment]
all the resource issues are now resolved

**Comment by czunker** [comment]
I gave it another try with the query pack, without success:
```
cnquery scan aws -f ../cnquery-packs/core/mondoo-aws-inventory.mql.yaml 
→ loaded configuration from /etc/opt/mondoo/mondoo.yml using source default
→ using service account credentials
! Scanning with local bundles will switch into --incognito mode by default. Your results will not be sent upstream.
→ discover related assets for 1 asset(s)

 AWS Account mondoo-dev (921877552404) ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────  n/a


Summary (1 assets)
==================

Target:     AWS Account mondoo-dev (921877552404)
asset doesn't support any querypacks
```

This might happen beauce of
```
    filters:
      - asset.platform == "aws"
```
in the query pack.

I tried to dig a bit deeper and got strange results:
```
cnquery run aws -c "asset{*}" 
→ loaded configuration from /etc/opt/mondoo/mondoo.yml using source default
rpc error: code = Unknown desc = cannot find connection type aws
asset: no data available

```
Doing the same again:
```
cnquery run aws -c "asset{*}" 
→ loaded configuration from /etc/opt/mondoo/mondoo.yml using source default
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0xdeac6d]

goroutine 30 [running]:
go.mondoo.com/cnquery/providers.(*Runtime).watchAndUpdate(0xc000322cc0, {0xc0006f3ae1, 0x5}, {0x0, 0x0}, {0xc0007a8708, 0x13}, {0x0?, 0xc0008e3338?})
	/home/christian/workspace/mondoo/github.com/cnquery/providers/runtime.go:267 +0xcd
go.mondoo.com/cnquery/providers.(*Runtime).WatchAndUpdate(0xc000099de0?, {0x1621dc0, 0xc000099dc0}, {0xc0007a8708, 0x13}, {0xc0002d20c0, 0x31}, 0xc00098c1e0)
	/home/christian/workspace/mondoo/github.com/cnquery/providers/runtime.go:247 +0xb2
go.mondoo.com/cnquery/llx.runResourceFunction(0xc000270460, 0xc0008bd890, 0xc000745a90, 0x20000000f)
	/home/christian/workspace/mondoo/github.com/cnquery/llx/builtin.go:717 +0x20a
go.mondoo.com/cnquery/llx.(*blockExecutor).runBoundFunction(0xc000270460, 0xc0008bd890, 0xc000745a90, 0xc0008e3520?)
	/home/christian/workspace/mondoo/github.com/cnquery/llx/builtin.go:791 +0xe5
go.mondoo.com/cnquery/llx.(*blockExecutor).runFunction(0xc000270460, 0xc000745a90, 0xc0008da1b0?)
	/home/christian/workspace/mondoo/github.com/cnquery/llx/llx.go:796 +0xd3

```

I think this all fails because of https://github.com/mondoohq/cnquery/issues/1486. We may need to fix this first to get valid results.

**Comment by vjeffrey** [comment]
oh interesting.
asset seems to work fine for shell..
```
~/go/src/go.mondoo.io/cnquery vj/aws-provider*
[5/09/23 08:28:02] ❯ go run apps/cnquery/cnquery.go shell aws
! using builtin provider for aws
→ loaded configuration from /Users/vj/.config/mondoo/mondoo.yml using source default
→ connected to Amazon Web Services
  ___ _ __   __ _ _   _  ___ _ __ _   _
 / __| '_ \ / _` | | | |/ _ \ '__| | | |
| (__| | | | (_| | |_| |  __/ |  | |_| |
 \___|_| |_|\__, |\__,_|\___|_|   \__, |
  mondoo™      |_|                |___/  interactive shell

cnquery> asset
asset: asset name="AWS Account mondoo-dev (921877552404)" platform="aws" version=""
```

my scan also worked:
```

Summary (1 assets)
==================

Target:     AWS Account mondoo-dev (921877552404)
Datapoints: 18
```

weirdness

**Comment by vjeffrey** [comment]
hmm, asset seems to work for me on run..
```
[5/09/23 08:31:56] ❯ go run apps/cnquery/cnquery.go run  aws -c 'asset { * }'
! using builtin provider for aws
→ loaded configuration from /Users/vj/.config/mondoo/mondoo.yml using source default
! using builtin provider for aws
asset: {
  kind: "api"
  build: ""
  version: ""
  title: "Amazon Web Services"
  arch: ""
  labels: {}
  platform: "aws"
  runtime: "aws"
  ids: [
    0: "//platformid.api.mondoo.app/runtime/aws/accounts/aws.account/921877552404"
  ]
  name: "AWS Account mondoo-dev (921877552404)"
  fqdn: ""
  family: []
}
```

**Comment by czunker** [comment]
As discussed, this is independent of the AWS provider. Will solve the asset thing in a separate PR.

**Review by czunker — APPROVED** [review]
Thanks @vjeffrey

**Comment by vjeffrey** [comment]
thanks for the review christian!

**Pull request merged**